### DISCUSSION TITLE
How do I properly override `__getattr__` to access dict keys as attributes?

### DISCUSSION THREAD

**Discussion by jtpavlock** [sqlalchemy/sqlalchemy#8511] [Usage Questions]
I'm trying to override `__getattr__` in SQLAlchemy so I can access certain dictionary fields using `object.field` syntax rather than having to do `object.fields["field"]`.

However, SQLAlchemy isn't recognizing changes to the field if I use the dot notation syntax. Accessing the field using normal `object.fields["field"]` however works as expected.

I've provided a code sample below.
```python
#!/usr/bin/env python3
import sqlalchemy
from sqlalchemy import Column, Integer, JSON
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.ext.mutable import MutableDict
from sqlalchemy.orm import sessionmaker

Session = sessionmaker()
Base = declarative_base()


class Track(Base):  # noqa: WPS230
    __tablename__ = "track"

    id = Column(Integer, primary_key=True)
    _fields = Column(MutableDict.as_mutable(JSON), default="{}")

    def __init__(self, id):
        self.id = id
        self._fields = {}
        self._fields["custom"] = "field"

    def __getattr__(self, name: str):
        """See if ``name`` is a custom field."""
        try:
            return self.__dict__["_fields"][name]
        except KeyError:
            raise AttributeError from None

    def __setattr__(self, name, value):
        """Set custom fields if a valid key."""
        if "_fields" in self.__dict__ and name in self.__dict__["_fields"]:
            self._fields[name] = value
        else:
            super().__setattr__(name, value)


engine = sqlalchemy.create_engine("sqlite:///:memory:")
Session.configure(bind=engine)
Base.metadata.create_all(engine)  # creates tables
session = Session()

track = Track(id=1)
session.add(track)
session.commit()

track.custom = "new"
# track._fields["custom"] = "new"  # works
assert track in session.dirty  # fails 
```

Is there any way to achieve the behavior I'm after?

(cross-posted from [stack overflow](https://stackoverflow.com/questions/73684997/how-do-i-properly-override-getattr-in-sqlalchemy-to-access-dict-keys-as-at) as I now realize this is the official place to get support)

**Comment by CaselIT** [comment]
Hi,

I had replied before bug the suggested solution was not working.
When you are running `track.custom` the track instance has no state, since it was expired by the commit, so the else route in `__setattr__` is teken. 

overall if you have a static list of keys in fields you could probably use that to validate the name in `__setattr__` otherwise something like this seems to work
```py
class Track(Base):  # noqa: WPS230
    __tablename__ = "track"

    id = Column(Integer, primary_key=True)
    _fields = Column(MutableDict.as_mutable(JSON), default="{}")

    def __init__(self, id):
        self.id = id
        self._fields = {}
        self._fields["custom"] = "field"

    def __getattr__(self, name: str):
        """See if ``name`` is a custom field."""
        try:
            return self.__dict__["_fields"][name]
        except KeyError:
            raise AttributeError(name) from None

    def __setattr__(self, name, value):
        """Set custom fields if a valid key."""
        if getattr(self, '_fields', None) and name in self._fields:
            self._fields[name] = value
        else:
            super().__setattr__(name, value)
```

**Reply by jtpavlock** [reply]
Thanks for the response. I apologize, but it seems my problem is slightly more nuanced than the sample code I provided suggests. Changing the tests to:

```python
assert track.custom # <-- fails
track.custom = "new" 
assert track in session.dirty
assert track.custom == "new"
```
Fails with the above solution. Also I seem to get an `UnmappedInstanceError` with the above solution that was not present before when testing it in my actual codebase. I unfortunately was unable to reproduce it in the test example, so I'm not sure how helpful you can be in that regard, but I'm attaching the log just in case you may have a better idea on why it could occur while I continue trying to recreate it in the sample code.

Edit: Unmarking the answer because while setting a static list work, generating the static list on Track initiation does not. 

[sa_field_unmapped_errror.txt](https://github.com/sqlalchemy/sqlalchemy/files/9552587/sa_field_unmapped_errror.txt)

**Reply by jtpavlock** [reply]
As you previously alluded to, setting a static list of custom fields works. 

```python
    def __getattr__(self, name: str):
        """See if ``name`` is a custom field."""
        if name in self.FIELDS:
            return self._fields[name]
        else:
            raise AttributeError from None

    def __setattr__(self, name, value):
        """Set custom fields if a valid key."""
        if name in self.FIELDS:
            self._fields[name] = value
        else:
            super().__setattr__(name, value)
```
It feels wrong to have a duplicate list of accessible fields between `_fields.keys()` and `FIELDS`, but I can't seem to get around the issues that come with trying to inspect `_fields`.

**Reply by CaselIT** [reply]
Right, since after a commit the status of `track` is empty. The orm will lazy load things, but the `__getattr__` interferes with it. You would probably need to call the class getattr from the getter to make it work as expected. 

A partial solution may be of setting expire on commit to false on the session https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=expire_on_commit#sqlalchemy.orm.Session.params.expire_on_commit

Maybe using a subclass of MappedDict that has getattr and setattr and then using `track.fields.custom` would be a more robust solution.

**Reply by jtpavlock** [reply]
Could you elaborate on what you mean by "You would probably need to call the class getattr from the getter to make it work as expected." ?

**Reply by CaselIT** [reply]
I'm not sure what I meant. Class getattr will not work, since it would return the column, not perform the lazy load. sorry
